### PR TITLE
Output ampq

### PR DESCRIPTION
**Pull request by pedromreis**
Hi!
Made some changes to the initial work by @sch3m4, mainly on the connection part:
- Changed from URL based connection to classic parameters
- Added some configurable parameters, needed to connect to my RabbitMQ platform
- Some cleanup on the try-except routines

This works perfectly with our RabittMQ. But I would love if someone else test it on different installations.

PR

### CODE REVIEW

File: intelmq/bots/outputs/amqptopic/output.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            content_type=self.CONTENT_TYPE, delivery_mode=self.DELIVERY_MODE)
+
+        try:
+            if not self._channel.basic_publish(exchange=self._exchange,
+                                               routing_key=self._routing_key,
+                                               body=event.to_json(),
+                                               properties=properties,
+                                               mandatory=True):
+                self.logger.error("Message sent but not confirmed")
+        except Exception:
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: First, only catch a specific subset of Exceptions you would expect in the basic_publish call

File: intelmq/bots/outputs/amqptopic/output.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            self.logger.error("Error publishing the message")
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: And second, we would like to have the original error, thus `logger.exceptin`

File: intelmq/bots/outputs/amqptopic/output.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        properties = pika.BasicProperties(
+            content_type=self.CONTENT_TYPE, delivery_mode=self.DELIVERY_MODE)
+
+        try:
+            if not self._channel.basic_publish(exchange=self._exchange,
+                                               routing_key=self._routing_key,
+                                               body=event.to_json(),
+                                               properties=properties,
+                                               mandatory=True):
+                self.logger.error("Message sent but not confirmed")
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: In the not confirmed state: Is this expected to happen from time to time and it's handled? Then it's a warning.
If it should not happen, but the message is not dropped, the error is ok in my eyes.
If it should not happen, but the message may be not received by the other end, the bot should fail IMHO (or at least by default, could be optional by configuration parameter)

File: REQUIREMENTS [hunk 4]
```diff
@@ -8,3 +8,5 @@ redis>=2.10.3
 requests>=2.7.0
 setuptools>=0.9
 wheel>=0.21
+pika>=0.10.0
```
Comment (sebix) [inline REQUIREMENTS]: Please put this into a bot-specific REQUIREMENTS, like this one: https://github.com/certtools/intelmq/blob/master/intelmq/bots/outputs/postgresql/REQUIREMENTS.txt

File: intelmq/bots/outputs/amqptopic/output.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            self._channel.confirm_delivery()
+
+    def process(self):
+        event = self.receive_message()
+
+        # verify | set connection
+        if None in (self._connection, self._channel):
+            self.connect_server()
+
+        properties = pika.BasicProperties(
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Why is this set every time?

File: intelmq/bots/outputs/amqptopic/output.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import sys
+
+from intelmq.lib.bot import Bot
+
+import pika
+
+
+class AMQPTopicBot(Bot):
+
+    CONTENT_TYPE = 'application/json'
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: They are only used once, why making a variable from it? Is it planned to extend the bot?

File: intelmq/bots/outputs/amqptopic/output.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    DELIVERY_MODE = 1
+
+    def init(self):
+        self._connection = None
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Leading underscores in python are private (by convention). Why do you use them?

**Comment by sebix** [comment]
I'd like to see a README explaining the bot (configuration), and possible use cases.

**Comment by pedromreis** [comment]
> I'd like to see a README explaining the bot (configuration), and possible use cases.

And you will :).

Thank you @sebix for your review and comments.
I'll look at it tomorrow.

**Comment by pedromreis** [comment]
@sebix 
Did you had time to review my last commits on this?

PR

**Comment by sebix** [comment]
No, sorry. Haven't had time to review and test yet.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+import sys
+
+import pika
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Have a look here: https://github.com/certtools/intelmq/blob/master/intelmq/bots/experts/asn_lookup/expert.py#L8

Using this method, the bot will fail at initialization time and thus log the error.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                                                                    self.connection_port, self.connection_vhost))
+            self.channel = self.connection.channel()
+            self.channel.exchange_declare(exchange=self.exchange, type=self.exchange_type, durable=self.durable)
+            self.channel.confirm_delivery()
+
+    def process(self):
+        ''' Stop the Bot if cannot connect to AMQP Server after the defined connection attempts '''
+
+        try:
+            if (self.connection.is_closed or self.channel.is_closed):
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Unnecessary brackets.

File: intelmq/bots/outputs/amqptopic/output.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                self.connect_server()
+        except AttributeError:
+            self.logger.exception('Bad configuration or server unavailable! Exiting...')
+            self.stop()
+
+        event = self.receive_message()
+
+        if (not self.keep_raw_field):
```
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Unnecessary brackets.

File: intelmq/bots/outputs/amqptopic/output.py (hunk 9 above)
Comment (sebix) [inline intelmq/bots/outputs/amqptopic/output.py]: Both values can also be `None`

**Comment by sebix** [comment]
Still missing docs.

With a running unconfigured server, I get this amount of output:

```
amqp: AMQP Connecting to 127.0.0.1:5672/fooobar 
amqp: AMQP connection to 127.0.0.1:5672/fooobar failled!!
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 384, in _handle_read
    data = self.socket.recv(self._buffer_size)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/amqptopic/output.py", line 42, in connect_server
    self.connection = pika.BlockingConnection(self.connection_parameters)
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/blocking_connection.py", line 339, in __init__
    self._process_io_for_connection_setup()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/blocking_connection.py", line 374, in _process_io_for_connection_setup
    self._open_error_result.is_ready)
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/blocking_connection.py", line 410, in _flush_output
    self._impl.ioloop.poll()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/select_connection.py", line 602, in poll
    self._process_fd_events(fd_event_map, write_only)
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/select_connection.py", line 443, in _process_fd_events
    handler(fileno, events, write_only=write_only)
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 364, in _handle_events
    self._handle_read()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 407, in _handle_read
    return self._handle_error(error)
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 338, in _handle_error
    self._handle_disconnect()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 288, in _handle_disconnect
    self._adapter_disconnect()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/select_connection.py", line 95, in _adapter_disconnect
    super(SelectConnection, self)._adapter_disconnect()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 154, in _adapter_disconnect
    self._check_state_on_disconnect()
  File "/usr/local/lib/python3.4/dist-packages/pika/adapters/base_connection.py", line 173, in _check_state_on_disconnect
    raise exceptions.ProbableAuthenticationError
pika.exceptions.ProbableAuthenticationError
amqp: Bot starts processings.
amqp: Loading source pipeline.
amqp: Loading source queue.
amqp: Source queue loaded amqp-output-queue.
amqp: Connected to source queue.
amqp: Loading destination pipeline.
amqp: Loading destination queues.
amqp: Destination queues loaded None.
amqp: Connected to destination queues.
amqp: Pipeline ready.
amqp: Start processing.
amqp: Bad configuration or server unavailable! Exiting...
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/amqptopic/output.py", line 60, in process
    if (self.connection.is_closed or self.channel.is_closed):
AttributeError: 'NoneType' object has no attribute 'is_closed'
amqp: Disconnecting from source pipeline.
amqp: Disconnecting from destination pipeline.
amqp: Bot stopped.
amqp: Exiting.
amqp: Bot stopped.
amqp: Exiting.
Exception ignored in: <bound method AMQPTopicBot.__del__ of <intelmq.bots.outputs.amqptopic.output.AMQPTopicBot object at 0x7f575b4bae48>>
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 188, in __del__
    self.stop()
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 202, in stop
    self.__terminate()
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 209, in __terminate
    exit(-1)
  File "/usr/lib/python3.4/_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)
SystemExit: -1
```
There's a `ConnectionResetError`, shouldn't this one be also catched? The AttributeError should be easy to fix. Not sure where the `Exception ignored ...` comes from.

**Comment by pedromreis** [comment]
> Not sure where the Exception ignored ... comes from." 

Can't find it either... need help on this...

@sebix Regarding the exception handling, add catching for the 'pika.exceptions.ProbableAuthenticationError'.

Regarding the AttributeError, no sure what needs to be done.

PR

**Comment by sebix** [comment]
I fixed the error handling and merged it.

**Pull request merged**